pkg/ext-proc/backend/provider.go

```diff
@@ -27,60 +29,73 @@ func NewProvider(pmc PodMetricsClient, datastore *K8sDatastore) *Provider {
 type Provider struct {
 	// key: Pod, value: *PodMetrics
 	podMetrics sync.Map
-	pmc        PodMetricsClient
-	datastore  *K8sDatastore
+	// key: Pod, value: *PodMetricsRefresher
+	podMetricsRefresher sync.Map
+	pmc                 PodMetricsClient
+	datastore           *K8sDatastore
 }
 
 type PodMetricsClient interface {
+	// FetchMetrics fetches metrics for the given pod.
+	// The returned PodMetrics and the existing one should not be the same object.
+	// Otherwise, there will be race.
 	FetchMetrics(ctx context.Context, pod Pod, existing *PodMetrics) (*PodMetrics, error)
 }
 
+func isPodMetricsStale(pm *PodMetrics) bool {
+	return time.Since(pm.UpdatedTime) > metricsValidityPeriod
+}
+
 func (p *Provider) AllPodMetrics() []*PodMetrics {
```

Comment:
I suggest having two helper functions to be super clear: AllFreshPodMetrics() and AllStalePodMetrics(). And if you need both, just combine the two lists.

Reply:
Done